docs/blog/20231031-savetheclocktower-modern-tree-sitter-part-4.md

```diff
@@ -0,0 +1,380 @@
+
+### Getting creative
+
+And it allows us to do some more complex things that weren’t possible before.
+
+TextMate’s system will let us indent or dedent _one level at a time_. But consider a `switch` statement:
+
+```js
+switch (foo) {
+	case "bar":
+		// one thing
+		break;
+	case "baz":
+		// another thing
+		break;
+	default:
+	// a third thing
+}
+```
+
+Between the second-to-last line and the last line, there’s a _two-level_ change in indentation. How can we express that?
+
+Here’s where our syntax tree comes in handy. Instead of describing our desired indentation level relative to the previous line, we can describe it relative to a line of our choosing:
+
+```scm
+; The closing brace of a switch statement's body should match the indentation
+; of the line where the switch statement starts.
+(switch_statement
+  body: (switch_body "}" @match)
+  (#set! indent.matchIndentOf parent.startPosition))
+
+; 'case' and 'default' need to be indented one level more than their containing
+; `switch`.
+(["case" "default"] @match
+  (#set! indent.matchIndentOf parent.parent.startPosition)
+  (#set! indent.offsetIndent 1))
+
+["case" "default"] @indent
+```
+
+Here we’re using a new capture called `@match`. It can do exactly what we just described by using a _node position descriptor_ (an idea we introduced [in the last installment](https://pulsar-edit.dev/blog/20231013-savetheclocktower-modern-tree-sitter-part-3.html#adjusting-by-node-position-descriptor)) to refer to an earlier line.
+
+In the first rule, we’re matching the closing `}` of a `switch` statement. We’re using a `#set!` predicate to describe the starting position of its `switch_body` parent. The `switch_body` starts on row 1, so `@match` understands this to mean “adjust the current line to match the indentation of row 1.” This will happen automatically when the user types the closing brace.
+
+In the second rule, we’re using similar logic. If we were typing the above `switch` statement, we’d find ourselves over-indented as we started typing on line 5. We’d want our editor to dedent that line once it saw that we were typing a new branch of the `switch` statement. So we can write another `@match` rule — still targeting the indentation level of the starting row of `switch_body` — but with an extra rule to _offset_ the indent by one level. In other words, we want to be indented _one level more_ than the indent level of row 1.
+
+The third rule is simpler: it’s how we ensure that the editor indents by one level after `case` and `default` statements.
+
+You might’ve had your hackles raised by this example. After all, there’s another school of thought on how to indent `switch` statements:
+
+```js
+switch (foo) {
+	case "bar":
+		// one thing
+		break;
+	case "baz":
+		// another thing
+		break;
+	default:
+	// a third thing
+}
+```
```

Comment:
Not sure if it is just my tired eyes but this looks the same as the previous switch statement example? Is this not meant to showcase the "alternate" style?

Reply:
That's weird. I'll fix it.

Reply:
OK, it seemed to be the work of a “helpful” pre-commit hook. That's annoying. Fixed!